### PR TITLE
Support the meta delete with tombstone to prevent dirty writes

### PR DESCRIPTION
**Pull request by drinkbeer**
Support the meta delete with tombstone to prevent dirty writes.

### CODE REVIEW

File: lib/dalli/client.rb [hunk 1]
```diff
@@ -273,10 +288,24 @@ def replace_cas(key, value, cas, ttl = nil, req_options = nil)
...
+    # Delete a key.
+    #
+    # `req_options` may include tombstone-mode keys, which leave a short-lived
+    # marker behind so concurrent readers (via `#get_with_status`) can tell a
+    # racing repopulate apart from a true miss:
+    # - `:invalidate` (Boolean) — mark the item stale instead of removing it.
+    # - `:tombstone_ttl` (Integer seconds) — how long the tombstone lives;
+    #   requires `:invalidate`. After this elapses, reads see `miss?`.
+    # - `:drop_value` (Boolean) — drop the stored value but keep the
+    #   tombstone marker, freeing memory while the tombstone lives.
```
Comment (mrattle) [inline lib/dalli/client.rb]: A bit of a nit but I don't think this comment is exactly true.  The value will be cleared and the page is marked to be re-occupied, but memory isn't actually _freed_ until the now-zero-value key is reallocated to another slab or it expires and gets evicted by the LRU.


```suggestion
    # - `:drop_value` (Boolean) — discard the stored value from the tombstone.
```

File: lib/dalli/protocol/meta/request_formatter.rb [hunk 2]
```diff
@@ -42,12 +42,18 @@ def self.meta_set(key:, value:, bitflags: nil, cas: nil, ttl: nil, mode: :set, b
           cmd << TERMINATOR
         end
 
-        def self.meta_delete(key:, cas: nil, ttl: nil, base64: false, quiet: false, p_token: nil, l_token: nil)
+        def self.meta_delete(key:, cas: nil, ttl: nil, base64: false, quiet: false, p_token: nil, l_token: nil,
+                             invalidate: false, tombstone_ttl: nil, drop_value: false)
+          raise ArgumentError, 'tombstone_ttl requires invalidate: true' if tombstone_ttl && !invalidate
```
Comment (mrattle) [inline lib/dalli/protocol/meta/request_formatter.rb:47]: I think it's worth noting that tombstoning via the `x` flag works differently than what we're thinking of when we compare against IDC's behaviour.  IDC left behind a standard value that had semantic meaning, passing `md foo x` (e.g. `drop_value: true` as the only tombstone option) simply drops the value.  At the memcached server level, the value is overwritten with `\r\n` - subsequent gets on that key (e.g. `mg foo v`) return a cache hit with a zero byte value, which I think is semantically different than a cached nil (the final result of the read after being received and parsed by Dalli is an empty string).  

Do we want to do that, or do we want to be more strict and require that we also set the invalidate flag here?

File: lib/dalli/client.rb (hunk 1 above)
Comment (drinkbeer) [reply lib/dalli/client.rb]: I think you are right. 👍

File: lib/dalli/protocol/meta/request_formatter.rb [hunk 3]
```diff
@@ -42,12 +42,18 @@ def self.meta_set(key:, value:, bitflags: nil, cas: nil, ttl: nil, mode: :set, b
...
+
           cmd = "md #{key}"
           cmd << ' b' if base64
           cmd << cas_string(cas)
           cmd << " T#{ttl}" if ttl
           cmd << ' q' if quiet
+          cmd << ' I' if invalidate
+          cmd << " T#{Integer(tombstone_ttl)}" if tombstone_ttl
```
Comment (mrattle) [inline lib/dalli/protocol/meta/request_formatter.rb:55]: Other methods sanitize incoming TTL via `TtlSanitizer` in `protocol/meta.rb`, I don't think we have that here so it would be good to add for consistency of behaviour across methods.

File: lib/dalli/protocol/meta/response_processor.rb [hunk 4]
```diff
@@ -67,6 +67,23 @@ def meta_get_without_value
...
+        # distinguish a tombstoned item (stale?) from a true miss (miss?).
+        # The value field may be empty when the tombstone was created with
+        # drop_value, which is intentional — callers branch on the
+        # predicates rather than nil-ness.
+        def meta_get_with_status
+          tokens = error_on_unexpected!([VA, EN, HD])
+          return ::Dalli::CacheResult.new(value: nil, miss: true) if tokens.first == EN
+
+          if tokens.first == VA
+            value = @value_marshaller.retrieve(read_data(tokens[1].to_i), bitflags_from_tokens(tokens))
```
Comment (mrattle) [inline lib/dalli/protocol/meta/response_processor.rb]: Looping back on the zero-byte return - I think it's possible that this might blow up without a check against that.  Dalli may try to unmarshal that zero byte value and return a deserialization error instead of an actual value to generate the intended `CacheResult` with.

**Review by mrattle — COMMENTED** [review]
A couple more comments on this.

File: lib/dalli/protocol/base.rb
Comment (oosidat) [inline lib/dalli/protocol/base.rb:1]: (Ref Line 69) Should this be rebased on main to include the changes from https://github.com/Shopify/dalli/commit/1636ef0b73af6c513206a839ecf70ce01358a956?

🐛 **Bug**: Main contains a fix (commit 1636ef0, merged via PR #71) that replaced the `$ERROR_INFO`-based ensure-clause gate with a local `request_local_completed` flag, specifically because `$!`/`$ERROR_INFO` remains non-nil inside an active `rescue` clause — causing `pipelined_get` to be torn down when `request` is called from inside a rescue. This branch was cut before that fix landed, so diffing against main shows the local-flag fix being removed and the predicate reverting to `close if $ERROR_INFO && @connection_manager.request_in_progress?`. The companion regression test (`'does not tear down a pipelined_get when called from inside a rescue clause'`) was likewise removed from `test/integration/test_fiber_concurrency.rb`. If this lands as-is, common cache-fallback patterns (`begin; ...; rescue; dc.get_multi(...) { ... }; end`) will silently return `{}` again. The new comment also misstates Ruby semantics — `$ERROR_INFO` is non-nil throughout an active rescue, not only while an exception is unwinding.

**Suggestion:** Rebase the branch onto current main so the local-flag fix and its regression test are preserved, then re-apply only the tombstone-related changes on top. The tombstone work is independent of the ensure-clause and the fiber-concurrency test, so a clean rebase should resolve it without conflicts.

File: lib/dalli/protocol/meta/response_processor.rb (hunk 4 above)
Comment (oosidat) [inline lib/dalli/protocol/meta/response_processor.rb]: Do we need to handle the case when `tokens[1].to_i` is zero?

🐛 **Potential Bug**: When a key is stored via the default path (`dc.set('tk', 'val')`), Marshal serialization is used and the FLAG_SERIALIZED bitflag is set. After `delete('tk', invalidate: true, drop_value: true)`, memcached preserves item flags but truncates the value to zero bytes. `meta_get_with_status` then calls `@value_marshaller.retrieve(read_data(0), bitflags_from_tokens(tokens))`, which routes the empty string into `Marshal.load('')`, raising ArgumentError that `ValueSerializer#retrieve` re-raises as `Dalli::UnmarshalError`. The current integration test only does `refute_equal 'should-be-dropped', result.value`, so it would surface as an unexpected exception rather than as a clear contract failure — and if memcached happens to strip flags on `x`, the bug stays latent until callers use non-string values.

**Suggestion:** Treat a zero-length VA body as an empty tombstone value rather than feeding it through the marshaller:
```suggestion
          if tokens.first == VA
            data_size = tokens[1].to_i
            value = data_size.zero? ? nil : @value_marshaller.retrieve(read_data(data_size), bitflags_from_tokens(tokens))
            @io_source.read(TERMINATOR.bytesize) if data_size.zero?
            ::Dalli::CacheResult.new(value: value, stale: stale_from_tokens(tokens))
```
Alternatively, confirm that memcached strips bitflags for dropped items and document that assumption in `meta_get_with_status`.

File: test/integration/test_tombstone.rb [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+
+        assert op_addset_succeeds(dc.set('tk', 'should-be-dropped'))
+
+        dc.delete('tk', invalidate: true, drop_value: true)
+        result = dc.get_with_status('tk')
+
+        assert_predicate result, :stale?
+        refute_predicate result, :miss?
+        # Value is dropped — empty string, not the original
+        refute_equal 'should-be-dropped', result.value
```
Comment (oosidat) [inline test/integration/test_tombstone.rb]: 💡 **Improvement**: `refute_equal 'should-be-dropped', result.value` passes for *any* value other than the original — including an unintended `nil`, a partial marshalled remnant, or an UnmarshalError-raised assertion failure. Since this test exists to verify the `x` flag's effect on the stored value, pinning the concrete expected shape would harden it against silent regressions and also surface the marshaller hazard discussed in the response_processor.rb finding.

**Suggestion:** 
```suggestion
        assert_equal '', result.value
```

**Review by oosidat — COMMENTED** [review]
Went through some of pair-review's suggestions, looks good overall, just wanted to highlight some potential edge cases.

---
_Review assisted by [pair-review](https://vault.shopify.io/posts/366774-devx-pair-review-is-now-available)_

File: test/integration/test_tombstone.rb (hunk 5 above)
Comment (oosidat) [reply test/integration/test_tombstone.rb]: Could also append the assertion to make it more watertight

File: lib/dalli/protocol/meta/request_formatter.rb (hunk 2 above)
Comment (drinkbeer) [reply lib/dalli/protocol/meta/request_formatter.rb:47]: Yeah. With `dc.delete("foo", drop_value: true)`, the command is `md foo x` which will clear the value payload but keep the key (value is 0 byte). In this case it is a hit because key still exist.

I read the protocol again, https://github.com/memcached/memcached/blob/master/doc/protocol.txt#L872

```
- x: removes the item value, but leaves the item.

This deletes the value off of an item (by replacing it with an empty value
item atomically). Combined with I this can leave what is effectively a
tombstone of a previous value.
```

Seems like the author of memcached intentionally allow the single `x`, but only when combined with `I` it can set the tombstone correctly. I prefer to keep the setting that allows single `x`, however, I updated the comments to explain what will happy with single `invalidate` (single `I`), single `drop_value` (single `x`) and combination of `invalidate` and `drop_value` (`I` + `x`).

File: lib/dalli/protocol/meta/request_formatter.rb (hunk 3 above)
Comment (drinkbeer) [reply lib/dalli/protocol/meta/request_formatter.rb:55]: Make sense. Addressed. Also added a test to cover the case of long tombstone_ttl.

File: lib/dalli/protocol/meta/response_processor.rb (hunk 4 above)
Comment (drinkbeer) [reply lib/dalli/protocol/meta/response_processor.rb]: Good call. We do want to intentionally support single `x` as memcached’s “drop value but leave item” behavior. I added integration coverage using a non-raw/serialized Dalli value to pin the zero-byte response behavior: after `drop_value: true` alone, `get_with_status` returns a non-stale hit with `value == ""`, and does not raise an unmarshal error.

File: lib/dalli/protocol/base.rb
Comment (drinkbeer) [reply lib/dalli/protocol/base.rb:1]: Good catch, thanks. This makes sense — we definitely should not revert the `request_local_completed` fix from 1636ef0.

I checked the rebased branch and the connection-cleanup fix is preserved: `Base#request` still uses the local `request_local_completed` flag rather than `$ERROR_INFO`, and the fiber-concurrency regression test (`does not tear down a pipelined_get when called from inside a rescue clause`) is still present. The tombstone changes are independent of that path.

I also verified the full suite locally with `bundle exec rake test`.

File: lib/dalli/protocol/meta/response_processor.rb (hunk 4 above)
Comment (drinkbeer) [reply lib/dalli/protocol/meta/response_processor.rb]: Good point. This is addressed with integration coverage for the zero-byte/single-`x` case.

The test now stores a non-raw/serialized Dalli value first, then deletes with `drop_value: true` only:

```ruby
dc.set("tk-x-only", { payload: "should-be-dropped" })
dc.delete("tk-x-only", drop_value: true)

result = dc.get_with_status("tk-x-only")
assert_predicate result, :hit?
refute_predicate result, :stale?
refute_predicate result, :miss?
assert_equal "", result.value
assert_equal "", dc.get("tk-x-only")
```

So if memcached returned a zero-byte value while preserving the serialized flag and Dalli tried to `Marshal.load("")`, this would fail with `Dalli::UnmarshalError`. On the memcached version we test against, `x` returns a zero-byte value that Dalli reads as `""`, which is the intended single-`x` behavior.

I did not apply the suggested `nil` special-case because single `x` should remain a cache hit with an empty string value, not a nil value. Full suite passes locally with `bundle exec rake test`.

File: test/integration/test_tombstone.rb (hunk 5 above)
Comment (drinkbeer) [reply test/integration/test_tombstone.rb]: Agreed — this has been addressed.

The assertion is now pinned to the concrete expected shape:

```ruby
assert_equal "", result.value
```

I also updated the `delete_multi(..., drop_value: true)` integration coverage to assert `""` explicitly instead of only checking that the original value changed. This makes the contract clear: `x`/`drop_value` leaves a zero-byte value that Dalli returns as an empty string.

Full suite passes locally with `bundle exec rake test`.

**Review by oosidat — APPROVED** [review]

File: lib/dalli/protocol/meta/response_processor.rb (hunk 4 above)
Comment (mrattle) [reply lib/dalli/protocol/meta/response_processor.rb]: What this test is confirming is that the bitflags are dropped along with the value, and that's fine since this is expected behaviour.

I did a quick LLM check against memcached source and that indicated that the server-side behaviour is that when the value is dropped the existing record isn't modified with an empty value, but completely overwritten as a new item with the same key, but no bitflags and a value of `\r\n`. Actually performing commands against a running memcached server locally showed that the server is returning that CRLF as a zero-byte value (empty string), which is what is returned to Dalli.  Since there are no bitflags returned, `ValueMarshaller` does not attempt to unmarshal and no deserialization error is raised.

**Review by mrattle — APPROVED** [review]
Looks good, I did some extra digging and tophatting to confirm the behaviour of the zero-byte value case and that seems fine without any additional handling.

File: lib/dalli/protocol/meta.rb [hunk 6]
```diff
@@ -112,13 +112,67 @@ def read_multi_req(keys, req_options = nil)
...
+          end
+
+          keys.each do |key|
+            results[key] = ::Dalli::CacheResult.new(value: nil, miss: true) unless results.key?(key)
+          end
+
+          unless attributes.frozen?
+            attributes['value_bytesize'] = total_value_bytesize
+            attributes['hit_count'] = results.count { |_key, result| result.hit? }
+            attributes['miss_count'] = results.count { |_key, result| result.miss? }
```
Comment (mrattle) [inline lib/dalli/protocol/meta.rb]: Two things here:
- we're treating stale hits as `miss=false` and `result.hit?` will return `true`. Do we want to be including them in the hit count when they're not actually returning usable data?  In the current setup where we simply do a hard delete the following reads would be treated as misses, so it feels like we're artificially inflating our hitrate with garbage data here.
- as this is written, `hit_count` + `miss_count` will sum to `keys.length`.  We're doing two passes of the keys here when we could just be doing a single pass for the misses and then subtracting that number from the length of the array of keys to get the number of hits for a minor performance optimization.  Further, this can probably also be wrapped into the `keys.each` block above.

File: lib/dalli/protocol/meta.rb [hunk 7]
```diff
@@ -112,13 +112,67 @@ def read_multi_req(keys, req_options = nil)
...
+      def read_multi_with_status_req(keys, req_options = nil)
+        results = SUPPORTS_CAPACITY ? Hash.new(nil, capacity: keys.size) : {}
+        optimized_for_raw = @value_marshaller.raw_by_default
+
+        total_value_bytesize = 0
+        @middlewares_stack.retrieve_req_pipeline('memcached.read_multi_with_status', { 'keys' => keys }) do |attributes|
+          routing_suffix = RequestFormatter.routing_tokens(**routing_token_kwargs(req_options))
+          post_get_req = optimized_for_raw ? "v k q#{routing_suffix}\r\n" : "v f k q#{routing_suffix}\r\n"
+          keys.each do |key|
+            @connection_manager.write("mg #{key} #{post_get_req}")
```
Comment (mrattle) [inline lib/dalli/protocol/meta.rb]: In this method we don't pass our key through`KeyRegularizer.encode` as we do in other methods.  This is carried through from the existing `read_multi_req` method, which has no documented reasoning for skipping that normalization and I think it was an error by ommission.

The end result is that you can potentially have a key with whitespace in it (which is forbidden by the server and will return an error), or you can have injected CRLF which can allow arbitrary commands to be sent via key injection.  I don't think that's actually a plausible vector for any use-cases that we currently have, but we still leverage this sanitization in (almost) every other method + the header parsing so it should probably be present here too.

File: lib/dalli/protocol/meta.rb [hunk 8]
```diff
@@ -181,6 +235,25 @@ def quiet_get_request(key, req_options = nil)
...
+        routing_kwargs = routing_token_kwargs(options)
+
+        @middlewares_stack.retrieve_req('memcached.get_with_status', { 'keys' => key }) do |attributes|
+          req = RequestFormatter.meta_get(key: encoded_key, value: true, base64: base64,
+                                          **routing_kwargs)
+          write(req)
+          @connection_manager.flush
+          result = response_processor.meta_get_with_status
+          unless attributes.frozen?
+            attributes['value_bytesize'] = result.value.nil? ? 0 : result.value.bytesize
```
Comment (mrattle) [inline lib/dalli/protocol/meta.rb]: Going to preface this one as another "bug" that exists elsewhere (in the existing `get` method) but does not seem to have any impact under our use conditions, so it can probably be safely ignored.

pair-review is calling out the potential for `result.value` to be a non-String type that doesn't respond to `.bytesize`, which would raise a `NoMethodError`.  

There are a set of conditions that need to be met for this to happen, and I don't think that we really encounter them:
- Need to be using OTel middleware so the above `attributes.frozen?` is false
- Need to be using Dalli directly, using Rails cache will protect against this
- Need to be caching non-String types without first serializing them

To handle this cleanly, you could capture the raw byte count inside `meta_get_with_status` (where the wire read already happens via `read_data`) and return it alongside the `CacheResult`:

```ruby
def meta_get_with_status
  tokens = error_on_unexpected!([VA, EN, HD])
  return [::Dalli::CacheResult.new(value: nil, miss: true), 0] if tokens.first == EN

  if tokens.first == VA
    raw_body = read_data(tokens[1].to_i)
    value = @value_marshaller.retrieve(raw_body, bitflags_from_tokens(tokens))
    [::Dalli::CacheResult.new(value: value, stale: stale_from_tokens(tokens)), raw_body.bytesize]
  else
    [::Dalli::CacheResult.new(value: nil, miss: true), 0]
  end
end
```

Then in `get_with_status`:

```ruby
result, raw_bytesize = response_processor.meta_get_with_status
unless attributes.frozen?
  attributes['value_bytesize'] = raw_bytesize
  attributes['hit_count']  = result.miss? ? 0 : 1
  attributes['miss_count'] = result.miss? ? 1 : 0
end
result
```

**Review by mrattle — COMMENTED** [review]
Commented on a couple of things that are extensions of latent behaviours (which don't seem to be blowing up anywhere as far as we know) and one telemetry-related thing

File: lib/dalli/protocol/meta.rb (hunk 6 above)
Comment (drinkbeer) [reply lib/dalli/protocol/meta.rb]: Good point — agreed that API/result semantics and hit-rate metrics should be separated here. A stale tombstone can still be `CacheResult#hit?` because memcached returned an item, but counting it as a metrics hit would inflate hit rate compared to the previous hard-delete behavior.

I changed the metrics semantics to:
- `hit_count`: fresh hits only
- `miss_count`: true misses + stale tombstones / other non-fresh results
- `stale_count`: stale tombstones, so we can observe them separately

For `read_multi_with_status`, the counts are computed while filling `results` rather than doing extra passes over the result hash.

File: lib/dalli/protocol/meta.rb (hunk 7 above)
Comment (drinkbeer) [reply lib/dalli/protocol/meta.rb]: Good catch — agreed. I updated this path to run keys through `KeyRegularizer.encode` and include the `b` flag when needed, rather than interpolating the raw key into the meta command. Created a helper method `write_read_multi_get` to process the keys.

I also applied the same fix to the existing `read_multi_req` path since it had the same omission. Response parsing now goes through `response_processor.key_from_tokens`, so base64 response keys are decoded back to the original key before namespace stripping / miss filling.

Added coverage for Unicode, whitespace, and CRLF-containing keys for both `get_multi` and `get_multi_with_status`.

Comment (drinkbeer) [reply lib/dalli/protocol/meta.rb]: Also added unit tests to cover the cases.

File: lib/dalli/protocol/meta.rb (hunk 8 above)
Comment (drinkbeer) [reply lib/dalli/protocol/meta.rb]: Agreed — this is a better place to measure the size. I updated `meta_get_with_status` to return both the `CacheResult` and the raw response body bytesize captured before deserialization, then `get_with_status` uses that raw bytesize for the OTel `value_bytesize` attribute.

I kept the updated stale metrics semantics from the earlier change (`hit_count` = fresh hits only, `miss_count` = non-fresh, `stale_count` separately reported).

Also added coverage for a non-String value through `get_with_status` with OTel enabled, which would have raised on `result.value.bytesize` before this change.
